### PR TITLE
Fix gradient checkpointing conflicts with HPU graphs and improve dataset preprocessing in Whisper example

### PR DESCRIPTION
**Pull request by gplutop7**
This PR resolves training instability and runtime errors caused by gradient checkpointing interacting with HPU graphs and caching in the Whisper speech recognition example.
Main fixes and improvements:

Gradient checkpointing fix

- Disabled use_cache=True when gradient checkpointing is enabled to prevent duplicated backward hooks on HPU.
- Automatically disables HPU graphs for inference during training to avoid runtime conflicts.
- Ensures safe, non-reentrant checkpointing for compatibility with PyTorch ≥ 2.2 and Transformers ≥ 4.55.

Example command that previously reproduced the issue:
`PT_HPU_LAZY_MODE=1 python ../gaudi_spawn.py \
--world_size 8 --use_mpi run_speech_recognition_seq2seq.py \
--model_name_or_path="openai/whisper-large" \
--dataset_name="regisss/common_voice_11_0_hi" \
--language="hindi" \
--task="transcribe" \
--train_split_name="train+validation" \
--eval_split_name="test" \
--gaudi_config_name="Habana/whisper" \
--max_steps="625" \
--output_dir="/tmp/whisper-large-hi" \
--per_device_train_batch_size="16" \
--per_device_eval_batch_size="2" \
--logging_steps="25" \
--learning_rate="1e-5" \
--generation_max_length="225" \
--preprocessing_num_workers="1" \
--max_duration_in_seconds="30" \
--text_column_name="sentence" \
--freeze_feature_encoder="False" \
--sdp_on_bf16 \
--bf16 \
--overwrite_output_dir \
--do_train \
--do_eval \
--predict_with_generate \
--use_habana \
--use_hpu_graphs_for_inference \
--label_features_max_length 128 \
--dataloader_num_workers 8 \
--gradient_checkpointing \
--throughput_warmup_steps 3`

### CODE REVIEW

File: optimum/habana/transformers/gradient_checkpointing.py [hunk 1]
```diff
@@ -339,16 +339,10 @@ def checkpoint(
...
-            "if use_reentrant is not passed. use_reentrant=False is "
-            "recommended, but if you need to preserve the current default "
-            "behavior, you can pass use_reentrant=True. Refer to docs for more "
-            "details on the differences between the two variants."
-        )
-        use_reentrant = True
+    if use_reentrant is None or use_reentrant:
+        # Transformers>=4.55 + PyTorch>=2.2 require non-reentrant checkpointing on HPU
+        # Reentrant mode conflicts with DDP on HPU (duplicate backward hooks)
+        use_reentrant = False
```
Comment (regisss) [inline optimum/habana/transformers/gradient_checkpointing.py:345]: Maybe we can add a warning stating that reentrant is not used and shortly explaining why?

Comment (gplutop7) [reply optimum/habana/transformers/gradient_checkpointing.py:345]: Warning added.
